megatron/model/gpt_model.py

```diff
@@ -338,36 +339,88 @@ def _logits_helper(embedding, lm_output):
+                r"\d+.mlp.dense_4h_to_h.bias",
+            ])
+        if args.normalization == 'layernorm':
+            patterns.extend([
+                r"\d+.input_layernorm.bias",
+                r"\d+.post_attention_layernorm.bias",
+                r"\d+.bias",
+            ])
+        return patterns
+
+    @staticmethod
+    def _get_row_parallel_param_patterns():
+        return [
+            r"\d+.mlp.dense_4h_to_h.weight",
+            r"\d+.self_attention.dense.weight",
+        ]
+
+    @staticmethod
+    def _get_swiglu_col_parallel_param_patterns():
+        args = get_args()
+        if not args.swiglu:
+            return []
+        patterns = [
+            r"\d+.mlp.dense_h_to_4h.weight",
+        ]
+        if args.add_bias_linear:
+            patterns.append(r"\d+.mlp.dense_h_to_4h.bias")
+        return patterns
+
+
     def universal_checkpoint_info(self):
         info = dict()
         if DS_UNIVERSAL_CHECKPOINT_INFO:
             # Vocabulary parameters (embeddings) that require special handling due to padding.
-            info[VOCABULARY_PARAMETER_PATTERNS] = [
-                r"tied_modules.embed.word_embeddings.weight"
-            ]
+            info[VOCABULARY_PARAMETER_PATTERNS] = self._get_vocab_param_patterns()
 
             # Replicated (shared) parameters on the pipeline dimension
-            info[PIPELINE_REPLICATED_PARAMETER_PATTERNS] = [
-                r"tied_modules.embed.word_embeddings.weight",
-                r"tied_modules.embed.position_embeddings.weight"
-            ]
+            info[PIPELINE_REPLICATED_PARAMETER_PATTERNS] = self._get_pp_replicated_param_patterns()
 
             # Parameter slices that should be averaged not concatenated.
-            info[TP_REPLICATED_PARAMETER_PATTERNS] = [
-                r"tied_modules.embed.position_embeddings.weight",
-                r"\d+.input_layernorm.weight",
-                r"\d+.input_layernorm.bias",
-                r"\d+.post_attention_layernorm.weight",
-                r"\d+.post_attention_layernorm.bias",
-                r"\d+.self_attention.dense.bias",
-                r"\d+.mlp.dense_4h_to_h.bias",
-                r"\d+.weight",
-                r"\d+.bias",
-            ]
+            info[TP_REPLICATED_PARAMETER_PATTERNS] = self._get_tp_replicated_param_patterns()
 
             # Parameter that are sliced on the row dimension
-            info[PARAMETER_WITH_ROW_PARALLELISM_PATTERNS] = [
-                r"\d+.mlp.dense_4h_to_h.weight",
-                r"\d+.self_attention.dense.weight",
-            ]
+            info[PARAMETER_WITH_ROW_PARALLELISM_PATTERNS] = self._get_row_parallel_param_patterns()
+
+            # SWIGLU parameters are first sliced on dim=0 to tp slices
+            # Then, each tp slice is chunked into 2 to create the linear layers L1, L2 used for silu(L1(x)) * L2(x))
+            info[PARAMETER_WITH_2_SUB_PARAMS_CAT_DIM_0] = self._get_swiglu_col_parallel_param_patterns()
```

Comment:
@mosheisland, I am wondering if we can generalize this for arbitrary sub_param or (nested params) structures, removing the hardcoded `n_param=2` and `cat_dim=0`.  My thought for doing this would be to make this entry a dict like the following, as opposed to a list: 

`dict=(n_params=2, cat_dim=0, pattern_list=self._get_swiglu_col_parallel_param_patterns())`

I am not claiming that the proposed dict and keys sufficiently covers all sub_param cases, but it could be a good starting point and easily extensible in future. 

What do you think?

@stas00, FYI

Reply:
it's easier to generalize when there are several different cases.

But perhaps at least the name in the dict should match the class method? i.e. either swiglu on both sides, or `PARAMETER_WITH_2_SUB_PARAMS_CAT_DIM_0` on both sides so that it could be extended down the road to other keys that aren't swiglu?

Reply:
@tjruwase, I was also thinking about such generalization. However, the swiglu is currently the only case where we have sub-parameters and it is hard to predict what-if-any will be the other cases and if such generalization will fit. But, if you think it worth it, I can add it. Let me know what you think.

Reply:
@mosheisland, I agree with you both to wait for additional use cases before exploring generalization. I also agree with @stas00 that to harmonize the `dict` key as either `swiglu` or `PARAMETER_WITH_2_SUB_PARAMS_CAT_DIM_0` on both sides.

Reply:
@stas00, @tjruwase, the reason why I used different names is that DeepSpeed should not be aware of a specific model or module in Megatron-DeepSpeed repo (or other repo that uses DeepSpeed). Therefore, I used `PARAMETER_WITH_2_SUB_PARAMS_CAT_DIM_0 `in DeepSpeed. However, in Megatron-DeepSpeed, the actual module is swiglu - so I explicitly used this name for clarity.
I think there are 2 options here:
1. Keep names as-is
2. Change Megatron-DeepSpeed to use `PARAMETER_WITH_2_SUB_PARAMS_CAT_DIM_0`

I am ok with both. Just let me know what you prefer.

Reply:
@misland-habana, thanks for the explanation. Your reasoning makes sense, especially about keeping the DeepSpeed-side name to be model-agnostic. 

I am actually fine with keeping it as-is. 

@stas00, your thoughts?

Reply:
The original way works for me, as we said more cases are needed before a generalization can be attempted. Thank you for the explanation of your reasoining, Moshe.